proposals/1113_ldap-ad-web-ui/README.md

```diff
@@ -0,0 +1,151 @@
+These clients should _not_ be trusted to handle a user's IDP credentials.
+Therefore, we need a mechanism for untrusted clients to acquire Pinniped's downstream tokens while
+leaving the IDP credential handling to the Pinniped supervisor.
+
+## Proposal
+Pinniped must provide a simple login screen in order to support UIs that wish
+to authenticate with the Pinniped Supervisor to gain access to a cluster without
+requiring each app to handle IDP credentials.
+
+### Goals and Non-goals
+
+#### Goals
+* Prevent OAuth clients, other than the Pinniped CLI, from providing credentials via the authorization request
+* Provide a minimal feature set (ie user id, password & submit button only)
+* Provide generalized error messaging for failed logins that do not expose sensitive information (i.e. we should say "invalid username or password"
+  but do not expose whether it's the username or password that's incorrect)
+* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)
+* Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)
+* Prevent LDAP injection attacks
+* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
+* Screens are accessible and friendly to screen readers
+* Screens are friendly to password managers
+
+#### Non-goals
+* A rich client (ie the use of javascript)
+* Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.
+* Branding & customization beyond the information listed in the goals used to identify the login screen belongs to Pinniped.
+* Supporting SSO integrations
+* Internationalization or localization. The CLI doesn't currently support this either.
+
+### Specification / How it Solves the Use Cases
+
+#### API Changes
+
+The supervisor must accept requests from other clients, as detailed
+in the (todo) proposal for dynamic client registration.
+When a client other than pinniped-cli makes an authorization endpoint request with `response_type=code` and their
+IDP is an LDAP or Active Directory IDP, the user will be redirected to the new login page.
+The login page should display the IDP name and indicate that it belongs to Pinniped.
+When a client other than the Pinniped CLI makes an authorization endpoint request with
+custom Username/Password headers, they should be rejected.
+
+The discovery metadata for LDAP/AD IDPS should indicate that they support a flow of `browser_authcode`.
+
+The pinniped cli should default to using the cli-based password flow, but when a tty is unavailable,
+it will open a browser to log in
+instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
+may wish to authenticate using the pinniped cli but without access to a terminal.
```

Comment:
Rather than making decision based on tty availability, the CLI should use the existing `--upstream-identity-provider-flow` flag. This has already been implemented in both `pinniped login oidc` and `pinniped get kubeconfig`. More details in thread on k8s Slack: https://kubernetes.slack.com/archives/C01BW364RJA/p1650306152560689

The existing tty available/unavailable in the CLI behavior should not need to change. It's there to correctly handle cases like where the user is using kubectl in a shell script or where the user is piping input into kubectl. We wouldn't want the browser to open in those cases.